internal/exec/stages/files/units_test.go

```diff
@@ -0,0 +1,68 @@
+// Copyright 2020 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package files
+
+import (
+	"testing"
+
+	"github.com/coreos/ignition/v2/config/shared/errors"
+	"github.com/coreos/ignition/v2/config/v3_1_experimental/types"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestParseInstanceUnit(t *testing.T) {
+	type in struct {
+		unit types.Unit
+	}
+	type out struct {
+		serviceInstance string
+		instance        string
+		parseErr        error
+	}
+	tests := []struct {
+		in  in
+		out out
+	}{
+		{in: in{types.Unit{Name: "echo@bar.service"}},
+			out: out{serviceInstance: "echo@.service", instance: "bar",
+				parseErr: nil},
+		},
+
+		{in: in{types.Unit{Name: "echo@foo.service"}},
+			out: out{serviceInstance: "echo@.service", instance: "foo",
+				parseErr: nil},
+		},
+		{in: in{types.Unit{Name: "echo@fooservice"}},
+			out: out{serviceInstance: "", instance: "",
+				parseErr: errors.ErrNoDotSuffix},
+		},
+		{in: in{types.Unit{Name: "echo@.service"}},
+			out: out{serviceInstance: "echo@.service", instance: "",
+				parseErr: nil},
+		},
+		{in: in{types.Unit{Name: "postgresql@9.3-main.service"}},
+			out: out{serviceInstance: "postgresql@.service", instance: "9.3-main",
+				parseErr: nil},
+		},
```

Comment:
What should happen with `echo.service`?

Reply:
It will be considered as a normal `systemd` unit service and not an instantiated service. There's a check we put here (https://github.com/coreos/ignition/pull/934/files#diff-5e9270cae143ce51a2ea351871b471fcR65) to ensure that we're calling `hasInstanceUnit` function only when a given `unit.name` contains `@`.

Reply:
It'd be more robust if the helper function at least failed in that case, and if we had a test for it.

Reply:
@bgilbert I think you are saying that you want `parseInstanceUnit()` to error if someone passes in a normal systemd unit (`echo.service`) and you'd like a test added here to verify it errors. Is that correct?

Reply:
Yup.